v4/export/dump.go

```diff
@@ -196,193 +188,386 @@ func (d *Dumper) Dump() (err error) {
 	return nil
 }
 
-func dumpDatabases(pCtx context.Context, conf *Config, connectPool *connectionsPool, writer Writer, rebuildConnFunc func(*sql.Conn) (*sql.Conn, error)) error {
+func (d *Dumper) dumpDatabases(conn *sql.Conn, writer *Writer) error {
+	conf := d.conf
 	allTables := conf.Tables
-	g, ctx := errgroup.WithContext(pCtx)
-	tableDataIRTotal := make([]TableDataIR, 0, len(allTables))
-	splitChunkStart := time.Now()
+	ctx, cancel := context.WithCancel(d.ctx)
+	defer cancel()
+	var writingGroup errgroup.Group
+	tableDataStartTime := time.Now()
 	for dbName, tables := range allTables {
-		createDatabaseSQL, err := ShowCreateDatabase(connectPool.extraConn(), dbName)
+		createDatabaseSQL, err := ShowCreateDatabase(conn, dbName)
 		if err != nil {
 			return err
 		}
-		if err := writer.WriteDatabaseMeta(ctx, dbName, createDatabaseSQL); err != nil {
+		err = writer.WriteDatabaseMeta(ctx, dbName, createDatabaseSQL)
+		if err != nil {
 			return err
 		}
 
 		if len(tables) == 0 {
 			continue
 		}
 		for _, table := range tables {
-			table := table
-			tableDataIRArray, err := dumpTable(ctx, conf, connectPool.extraConn(), dbName, table, writer)
+			meta, err := dumpTableMeta(conf, conn, dbName, table)
+			if err != nil {
+				return err
+			}
+
+			if table.Type == TableTypeView {
+				err = writer.WriteViewMeta(ctx, dbName, table.Name, meta.ShowCreateTable(), meta.ShowCreateView())
+			} else {
+				err = writer.WriteTableMeta(ctx, dbName, table.Name, meta.ShowCreateTable())
+			}
+			if err != nil {
+				return err
+			}
+
+			tableIRStream := make(chan TableDataIR, defaultDumpThreads)
+			writingGroup.Go(func() error {
+				return writer.WriteTableData(ctx, meta, tableIRStream)
```

Comment:
We should use a global `tableIRStream` for `writingGroup` and start `goroutine` for each chunk (each mysql connection) like before. I think a good solution for `writingGroup` is to start `Threads`*worker `goroutines`. Each goroutine will use the same connection and try to get a chunk from the global `tableIRStream` and do the dumping work.

Imagining that we have a huge table. We will have only one `tableIRStream`, and in `writer.WriteTableData` the dump progress for each chunk is serialized not parallelized. The performance will sharply regress.